src/firefly/js/visualize/draw/ShapeDataObj.js

```diff
@@ -383,67 +622,96 @@ function drawText(drawObj, drawTextAry, plot, inPt, drawParams) {
         //TODO: better solution would be to adapt text color with background
 
         // in case shape 'text' defines the color of its own
-        var color = drawObj.sType === ShapeType.Text&&has(drawObj, 'color') ? drawObj.color : 'black';
+        var color = has(drawObj, 'color') ? drawObj.color : 'black';
         DrawUtil.drawText(drawTextAry, text, x, y, color, renderOptions,
                 fontName+FONT_FALLBACK, fontSize, fontWeight, fontStyle);
+        drawObj.textImgLoc = plot.getImageCoords(makeViewPortPt(x, y));
     }
 }
 
 /**
  *
  * @param drawObj
  * @param ctx
+ * @param drawTextAry
  * @param plot
  * @param drawParams
  * @param onlyAddToPath
  */
 function drawRectangle(drawObj, ctx, drawTextAry,  plot, drawParams, onlyAddToPath) {
-    var {pts, text, width, height, renderOptions, angle, angleUnit}= drawObj;
-    var {color, lineWidth, style, textLoc, unitType}= drawParams;
+    var {pts, text, width, height, renderOptions, angle = 0.0, angleUnit, isCenter = false, isOnWorld = false}= drawObj;
```

Comment:
the drawRectangle function is huge.  I am scared it will be hard to maintain.  Is there anything you can do to break it up?